### PR TITLE
Simplify widget render

### LINKED ISSUE #316: Break out render() code
Break out `render()` code into more manageable chunks, specifically in the following ways:

   - For widgets that generate more DOM, move sensible chunks into separate functions, e.g. `renderControls()` for the buttons in ComboBox, or `renderOptions` in Listbox.
   - For widgets that render children, e.g. TabController, return `this.children` in a separate function. Just as an example, this would make it much easier to add custom loading content.
   - Split "state" classes into a separate function (`getModifierClasses`?) for all widgets that could potentially support multiple visual states, e.g. `loading` or `dirty` on a tab button.
   - If a widget uses an icon, generate the icon separately. Swapping out icons or using something other than a font icon for a visual symbol seems like a fairly common modification, e.g. changing left/right month arrows in Calendar for text, or tiny month preview images.
   - Widgets that calculate inline styles (right now only Slider, I think) should do so in a separate fn. This would allow a triangular volume slider to adjust the height of the thumb as it changes, for example.

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Resolves #316 

Makes widgets more easily extensible by simplifying the render code, particularly addressing these points from the issue:
- break out large chunks of rendered dom
- state classes are returned in a function
- icons are broken out
- inline styles are calculated separately

### CODE REVIEW

File: src/button/Button.ts [hunk 1]
```diff
@@ -70,6 +70,30 @@ export default class Button extends ButtonBase<ButtonProperties> {
 	private _onTouchEnd (event: TouchEvent) { this.properties.onTouchEnd && this.properties.onTouchEnd(event); }
 	private _onTouchCancel (event: TouchEvent) { this.properties.onTouchCancel && this.properties.onTouchCancel(event); }
 
+	protected getContent() {
```
Comment (bitpshr) [inline src/button/Button.ts]: It may be nice to embrace ES getters if no arguments are passed to these:

```ts
get content() {
    return this.children;
}
```

File: src/calendar/tests/unit/Calendar.ts [hunk 2]
```diff
@@ -20,11 +20,9 @@ const compareId = compareProperty((value: any) => {
 	return typeof value === 'string';
 });
 
-let dateIndex = -1;
 const expectedDateCell = function(widget: any, date: number, active: boolean) {
-	dateIndex++;
 	return w(CalendarCell, {
-		key: `date-${dateIndex}`,
+		key: `date-${date}${active ? '' : '-dimmed'}`,
```
Comment (bitpshr) [inline src/calendar/tests/unit/Calendar.ts]: Why was this change necessary?

File: src/calendar/Calendar.ts [hunk 3]
```diff
@@ -16,7 +16,6 @@ import * as iconCss from '../common/styles/icons.m.css';
  *
  * Properties that can be set on a Calendar component
  *
- * @property CustomDateCell    Custom widget constructor for the date cell. Should use CalendarCell as a base.
```
Comment (bitpshr) [inline src/calendar/Calendar.ts]: Did you remove sub-widgets as part of this PR as well?

File: src/button/Button.ts (hunk 1 above)
Comment (tomdye) [reply src/button/Button.ts]: i'd say this is overkill @bitpshr especially if the `getter` ends up doing more than simply getting.

File: src/calendar/Calendar.ts [hunk 4]
```diff
@@ -321,38 +320,65 @@ export default class Calendar extends CalendarBase<CalendarProperties> {
...
+			onRequestMonthChange: (requestMonth: number) => {
+				onMonthChange && onMonthChange(requestMonth);
+			},
+			onRequestYearChange: (requestYear: number) => {
+				onYearChange && onYearChange(requestYear);
+			}
+		});
+	}
+
+	protected renderPagingButtonContent(type: 'next' | 'previous') {
```
Comment (tomdye) [inline src/calendar/Calendar.ts]: `next` / `previous` would be better as a string enum

File: src/calendar/DatePicker.ts [hunk 5]
```diff
@@ -191,7 +191,38 @@ export default class DatePicker extends DatePickerBase<DatePickerProperties> {
 		onPopupChange && onPopupChange(this._getPopupState());
 	}
 
-	private _renderMonthRadios() {
+	protected renderControlsTrigger(type: 'month' | 'year'): DNode {
```
Comment (tomdye) [inline src/calendar/DatePicker.ts]: string enum

File: src/calendar/DatePicker.ts [hunk 6]
```diff
@@ -215,7 +246,20 @@ export default class DatePicker extends DatePickerBase<DatePickerProperties> {
 		]));
 	}
 
-	private _renderYearRadios() {
+	protected renderPagingButtonContent(type: 'next' | 'previous') {
```
Comment (tomdye) [inline src/calendar/DatePicker.ts]: string enum

File: src/button/Button.ts (hunk 1 above)
Comment (bitpshr) [reply src/button/Button.ts]: Fair.

File: src/calendar/Calendar.ts (hunk 3 above)
Comment (smhigley) [reply src/calendar/Calendar.ts]: The problem is sub-widgets make it much easier to handle data binding on event handlers (e.g. `onClick` on a specific date cell returns the correct date and disabled state). Based on talking with Ant, I think the compromise reached was to leave in sub-widgets, but try not to expose them to the user

File: src/calendar/tests/unit/Calendar.ts (hunk 2 above)
Comment (smhigley) [reply src/calendar/tests/unit/Calendar.ts]: when I moved rendering the actual date cell into its own function (`renderDateCell`), I didn't have access to the index any more, and it seemed overkill to pass it in just for the key, when I could also generate a unique key based on the date and currentMonth.

File: src/calendar/Calendar.ts (hunk 3 above)
Comment (smhigley) [reply src/calendar/Calendar.ts]: I've been wondering if it makes sense to remove the typedoc for CalendarCell. It's marginally useful to developers wanting to extend CalendarCell, but I definitely don't want it showing up in Dojo documentation

File: src/timepicker/TimePicker.ts [hunk 7]
```diff
@@ -166,41 +166,50 @@ export const TimePickerBase = ThemeableMixin(WidgetBase);
...
+	}
+
+	private _getOptionLabel(value: TimeUnits) {
+		const { getOptionLabel } = this.properties;
+		const units = parseUnits(value);
+		return getOptionLabel ? getOptionLabel(units) : this._formatUnits(units);
+	}
+
+	private _onNativeBlur(event: FocusEvent) {
+		this.properties.onBlur && this.properties.onBlur((<HTMLInputElement> event.target).value);
```
Comment (pottedmeat) [inline src/timepicker/TimePicker.ts]: You could add this interface and use this instead of `FocusEvent` to clean things up a little.

```ts
interface FocusInputEvent extends FocusEvent {
	target: HTMLInputElement;
}
```

File: src/titlepane/TitlePane.ts [hunk 8]
```diff
@@ -108,36 +132,19 @@ export default class TitlePane extends TitlePaneBase<TitlePaneProperties> {
...
-					closeable ? css.closeableFixed : null,
-					css.titleFixed
-				),
-				onclick: this._onTitleClick,
-				onkeyup: this._onTitleKeyUp,
+				classes: this.classes(css.title, ...this.getModifierClasses()).fixed(css.titleFixed, ...this.getFixedModifierClasses()),
 				role: 'heading'
 			}, [
-				v('div', {
+				v('button', {
```
Comment (pottedmeat) [inline src/titlepane/TitlePane.ts]: Why did this change to a button?

File: src/titlepane/TitlePane.ts [hunk 9]
```diff
@@ -108,36 +132,19 @@ export default class TitlePane extends TitlePaneBase<TitlePaneProperties> {
...
-					'aria-disabled': closeable ? null : 'true',
 					'aria-expanded': String(open),
+					disabled: !closeable,
 					classes: this.classes(css.titleButton),
 					id: this._titleId,
-					role: 'button',
-					tabIndex: closeable ? 0 : -1
+					onclick: this._onTitleClick
 				}, [
-					v('i', {
```
Comment (pottedmeat) [inline src/titlepane/TitlePane.ts]: Why isn't this displayed anymore?

File: src/calendar/DatePicker.ts [hunk 10]
```diff
@@ -191,7 +207,38 @@ export default class DatePicker extends DatePickerBase<DatePickerProperties> {
...
+		const onclick = type === Controls.month ? this._onMonthButtonClick : this._onYearButtonClick;
+
+		return v('button', {
+			key: `${type}-button`,
+			'aria-controls': `${this._idBase}_${type}_dialog`,
+			'aria-expanded': `${open}`,
+			'aria-haspopup': 'true',
+			id: `${this._idBase}_${type}_button`,
+			classes: this.classes(
+				(css as any)[`${type}Trigger`],
```
[the remark this answers is not in the document]
Comment (smhigley) [reply src/calendar/DatePicker.ts]: yeah, I chose concise code over css types in this particular case, but I was kinda ambivalent about it.

File: src/titlepane/TitlePane.ts (hunk 8 above)
Comment (smhigley) [reply src/titlepane/TitlePane.ts]: I always thought it should be a button, since one of the basic a11y principles in our widget design is to use native elements whenever possible. I just took this widget pattern PR as a good place to change it. It also simplifies the code a ton, since we no longer need to handle focus and key events manually.

File: src/titlepane/TitlePane.ts (hunk 9 above)
Comment (smhigley) [reply src/titlepane/TitlePane.ts]: it is, I just moved it to `this.renderExpandIcon()`

File: src/calendar/DatePicker.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (pottedmeat) [reply src/calendar/DatePicker.ts]: The closest thing I found is:

```ts
const ControlTrigger: { [key: string]: 'monthTrigger' | 'yearTrigger' } = {
	[Controls.month]: 'monthTrigger',
	[Controls.year]: 'yearTrigger'
};

const ControlTriggerActive: { [key: string]: 'monthTriggerActive' | 'yearTriggerActive' } = {
	[Controls.month]: 'monthTriggerActive',
	[Controls.year]: 'yearTriggerActive'
};
```

```ts
css[ControlTrigger[type]],
open ? css[ControlTriggerActive[type]] : null
```

[the remark this answers is not in the document]
Comment (pottedmeat) [reply src/calendar/DatePicker.ts]: ^^ This still misses type safety on the key

[the remark this answers is not in the document]
Comment (pottedmeat) [reply src/calendar/DatePicker.ts]: Something like this adds type safety and minimal code but does require a lot of manually typed strings.

```ts
function controlClass(control: Controls, suffix: 'Trigger' | 'TriggerActive'): 'monthTrigger' | 'monthTriggerActive' | 'yearTrigger' | 'yearTriggerActive' {
	return `${control}${suffix}` as any;
}
```
